Add pitch and roll to TS helicopters

pchote opened the pull request:
Aircraft in TS would pitch forward by 30 degrees while flying, and roll to an angle of 20 degrees while turning.
This was effectively purely visual effect, and we have converted enough of the facing code to enable this for everything except the carryall.

https://i.imgur.com/Er7j4Bp.mp4

~Depends on #18228~

~I have taken the liberty of adjusting the pitch and roll values to look better with the current speed configuration on the different aircraft. I have also taken the approach from #17796 to define the roll as a factor of the turn amount, which looks more natural for aircraft that have different turn vs idle turn speeds.~

penev92 commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -208,21 +217,48 @@ public class Aircraft : PausableConditionalTrait<AircraftInfo>, ITick, ISync, IF
+
 		[Sync]
-		public WAngle Facing;
+		public WAngle Facing
```
Why is only this property synced?

pchote replied:
Pitch and roll have no impact on the gameplay simulation, so there is no need benefit to the performance price of syncing them.

Inq8 commented:
The Banshee does a strange-looking bob about 75% of the way around the turn in that mp4 example.

Is that just a recording blip?

pchote commented:
I think it may be passing off corner of a cliff briefly. There is another minor issue visible where it waggles its wings a bit - i may want to add a check to not roll if the delta-facing is less than one step.

tovl commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -209,7 +209,26 @@ public class Aircraft : PausableConditionalTrait<AircraftInfo>, ITick, ISync, IF
+
+				return orientation;
+			}
+		}
```
This seems like an unnecessarily roundabout way of doing things. Why not just do something like:

```C#
		public WAngle Facing
		{
			get { return Orientation.Yaw; }
			set { Orientation = Orientation.WithYaw(value); }
		}

		public WRot Orientation { get; set; }
```

and let `Fly` set `aircraft.Orientation` directly?

tovl commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -56,6 +56,15 @@ public class AircraftInfo : PausableConditionalTraitInfo, IPositionableInfo, IFa
+		[Desc("Body pitch to use when flying forwards (only relevant for voxel aircraft).")]
+		public readonly WAngle FlyPitch = WAngle.Zero;
+
+		public readonly WAngle PitchSpeed = WAngle.Zero;
```
Missing description

tovl commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -56,6 +56,15 @@ public class AircraftInfo : PausableConditionalTraitInfo, IPositionableInfo, IFa
+
+		[Desc("Amount to roll the body as a multiplier of the facing delta (only relevant for voxel aircraft).")]
+		public readonly int RollMultiplier = 0;
+		public readonly WAngle RollSpeed = WAngle.Zero;
```
Missing description

tovl commented:
It might be nice to replace `FlyPitch` with a `PitchMultiplier` as well. Letting it depend on the aircraft speed. That would make sure we already have a hook for when we implement aircraft acceleration.

pchote commented:
I was actually planning on going the other way, removing `RollMultiplier` and adding back `TurnRoll` and `IdleTurnRoll`. Multipliers have the same limited resolution issues that we are going to so much effort to remove in the facing rework.

tovl commented:
Hmm, I see what you mean, but I think the issue there is that the multiplier should just be defined as a percentage instead. (actually it would make sense to have fractions defined as per 1024 instead of per 100, since that is our standard resolution for every other measure, but I think there is no precedent for that)

abcdefg30 commented:
> (actually it would make sense to have fractions defined as per 1024 instead of per 100, since that is our standard resolution for every other measure, but I think there is no precedent for that)

Won't that get a problem with all of our other modifiers? I.e. why should a health or damage modifier have a base of 1024, or why should different modifiers behave so differently? (It can be done, but sounds quite unintuitive to me.)

tovl commented:
That is exactly what I meant by "there is no precedent for that"...

pchote committed:
Add Orientation getter to IFacing.

pchote commented:
Updated.

pchote commented:
Rebased.

pchote commented:
Updated: made some additional tweaks to improve the behaviour of TS aircraft:
* Removed Sliding - TS's actual behaviour was somewhere between our CanSlide: true and CanSlide: false, having it disabled feels better until we can implement it properly.
* Allow Banshee and Orca Bomber to hover - this is a better fit to the original game, and provides a testcase for our AttackType: Strafe.
* Increased movement speed to match the relative speed of ground vehicles (~14.2 x the rules.ini value) - TS is in desperate need of something like #18051 to fix the speed inconsistencies, but until that happens the newer values feel much better especially on the orca bomber.

Smittytron commented:
Roll looks good. Is it possible to slow down the pitch rate? The rate of pitching forward and backward looks a bit abrupt imo, especially on the orca transport.

pchote commented:
The fast pitch rate was intentional because IMO it looks wrong to have the pitch slowly adjusting while the unit is otherwise stationary. I could perhaps split out two separate pitch rates for active vs stationary changes so it pitches forward slower than it returns after stopping?

tovl commented:
> Removed Sliding - TS's actual behaviour was somewhere between our CanSlide: true and CanSlide: false, having it disabled feels better until we can implement it properly.

The sliding behaviour in TS was actually exactly the same as the sliding behaviour in earlier games, so it  doesn't seem justified to change this in TS only. If we want to fix this, we should just rework the sliding behaviour itself. The original games achieved this halfway behaviour by giving aircraft two facings: one for the sprite/voxel orientation and one for the flight direction. By giving these two different turnspeeds the original behaviour can be exactly recreated. Our current sliding behaviour is then equivalent to having the turnspeed for the flight direction maximized.

> Allow Banshee and Orca Bomber to hover - this is a better fit to the original game, and provides a testcase for our AttackType: Strafe.

In the original game, Banshees and Orca bombers never hovered in place, unless they were landing. We already split out the VTOL behaviour so this change makes no sense.

pchote commented:
The same was true for Orca and Harpy - the only difference is that those two could hover while attacking, whereas the other two used a strafing/bombing run style attack. The fact that the orcab/scrin can hover is demonstrated by their ability to land, so IMO it makes no sense for them to circle when idle.

tovl commented:
In that case it makes just as little sense for D2K carryalls to circle when idle, but they do.

pchote commented:
In D2k's case it is for consistency with the original game, and the same applies to an extent here:
* The orcab artwork is clearly helicopter/non-bomber-orca-like, so there is no reason for it to circle if the regular orca doesn't
* The circling interacts badly with the no-move-into-shroud logic, either allowing you to slowly walk them across the map (with the current reveal behaviour) or lose them under the shroud (if we properly remove the revealing).

I didn't think this change would be controversial, but if it is then my preference would be to reintroduce the full OG behaviour when they land while idle - as frustrating as that logic is - and defer the arguments until we have a properly playable mod where balancing of everything can be properly considered together. This would need to apply to the carryall too for consistency, which is an even bigger shame considering how well that is working now.

Smittytron commented:
> The fast pitch rate was intentional because IMO it looks wrong to have the pitch slowly adjusting while the unit is otherwise stationary.

I can't think of a happy middle ground without making things overly complicated, so I'm fine with keeping pitch as you have it now.

tovl commented:
For what it's worth, I do not object to reconsidering the behaviour of these craft on principle. Indeed, there is an argument to be made for the orca bomber (but the same argument can be made for the D2K carryall, so...). I do object to you claiming this is somehow more consistent with the original game than what we have now, when it is quite clear that the original game provides no precedent whatsoever on this, one way or the other, since in the original they just land on idle (which is awful). In fact, if we consider the later games, there is plenty of precedent against it (Firehawks, Vertigo bombers, Storm Riders all circle when idle while being VTOLs). As it stands, this is an unmotivated change, that is not an objective improvement and is a scope creep from what the PR claims to be. I would advise to just open a new issue to gather opinions about such a change, and leave it out of this PR.

> The circling interacts badly with the no-move-into-shroud logic, either allowing you to slowly walk them across the map (with the current reveal behaviour) or lose them under the shroud (if we properly remove the revealing).

This will be an issue with any mod that combines fixed-wing aircraft with no-move-into-shroud logic, which seems not far-fetched at all. This is a problem that we need to deal with head-on, not sidestep, provided we want to support no-move-into-shroud logic at all. (But this is not something that needs to be addressed in this PR).

> I didn't think this change would be controversial, but if it is then my preference would be to reintroduce the full OG behaviour when they land while idle - as frustrating as that logic is - and defer the arguments until we have a properly playable mod where balancing of everything can be properly considered together. This would need to apply to the carryall too for consistency, which is an even bigger shame considering how well that is working now.

This is just childish. Please don't stoop to threatening to flip the table.

Regarding the pitch rate: Ideally, the aircraft would adjust its pitch before stopping, not after. I don't know how simple it would be to incorporate that here, otherwise I would be happy to take a look at this myself in a follow up. It would tie in nicely with acceleration/deceleration.

pchote commented:
I didn't intend for that comment as a hammer to force my way... i really do think that when we put ourselves into a corner with no clear escape that the sensible option is to fall back to the original logic (which is just a yaml tweak here) - we have time to consider this again before we release anything to the public.

How about a compromise where we put the orcab back to VTOL, keep the banshee circling, but then fix the vision revealing so that it disappears under the shroud (maybe this isn't as bad as I feared)?

tovl commented:
If you insist on doing this here, then making just the orca bomber hovering on idle but keeping the banshee circling seems like a reasonable compromise, considering the visuals. 

As for the vision changes: I just checked and in the old game aircraft did reveal shroud and not just when landing either. It was totally possible to slowly walk across the map this way as well (or quickly if you already have a spot revealed on the other side of the map). That the aircraft didn't circle on idle didn't really matter here, as the vision was more than one cell anyway. The same applied to subterranean units. So if you want consistency with the original games, the current behaviour is already there.

pchote commented:
Ok. Restored circling on the banshee and left the vision alone - we still make it a lot easier even for non-banshees, but i suspect this may end up falling under the banner of QOL improvements...

deleted-user-1 reviewed (commented):
Please make banshees hover again. I first objected them hovering too, but after comparing the original and our hovering/circling banshee I think the hovering is more in line with the original and looks overall better. At least there needs to be an IdleTurnSpeed to make it look less like meme.

> I do object to you claiming this is somehow more consistent with the original game than what we have now, when it is quite clear that the original game provides no precedent whatsoever on this, one way or the other, since in the original they just land on idle (which is awful).

@tovl Hovering banshees are more in line with the original because the circling is non-existent in the original but VTOL is. This gives direction for ways to fix autolanding. My impression after playing both games and setups is that adding circling on idle changes the look and feel of the game way more than hovering does. Let's please try it for TS the other way than it is in RA. I can't imagine that it would look good when many units circle in an area with height differences and for gameplay and unit control hovering will be better.

> I could perhaps split out two separate pitch rates for active vs stationary changes so it pitches forward slower than it returns after stopping?

I think that would work, the original had the same abrupt pitch like here so that would be a nice improvement.

> If we want to fix this, we should just rework the sliding behaviour itself. The original games achieved this halfway behaviour by giving aircraft two facings: one for the sprite/voxel orientation and one for the flight direction.

That sounds like how it looks like in the original. IMO OpenRA sliding behavior was already close and good enough. Now Orca and Harpy can not be micro-ed from cell to cell without moving autonomously in unpredictable ways. I don't see the need to make it worse than it could already be tbh.

I would probably use different configurations for roll, speed and pitch and make it overall more subtle. I am bad a spotting visual bugs, testing in TS and RA worked fine.

penev92 commented:
A PR that is supposed to be about a basic engine feature should really not get bogged down over gameplay politics! (aircraft pitching/tilting/rolling vs whether a specific unit should hover, circle, land or return to base and how each of those is good or bad for gameplay)

pchote commented:
> I don't see the need to make it worse than it could already be tbh.

Sliding breaks the banshee and orca bomber - see #17935.

deleted-user-1 commented:
> Sliding breaks the banshee and orca bomber - see #17935.

I might be missing something but I thought to only add it for orca and harpy. In the original banshees can't slide to the sides (orca and harpy can, don't know about orca bomber). My idea was that all aircraft become hovering vtols and banshee and orca bomber keep their non-sliding-plane-alike behavior (however as penev noted this should not block progress).

dnqbob commented:
> > Sliding breaks the banshee and orca bomber - see #17935.
> 
> I might be missing something but I thought to only add it for orca and harpy. In the original banshees can't slide to the sides (orca and harpy can, don't know about orca bomber). My idea was that all aircraft become hovering vtols and banshee and orca bomber keep their non-sliding-plane-alike behavior (however as penev noted this should not block progress).

I think it is the engine limit make aircrafts not slide/cycling in original TS.  because in CG Banshee slides well.

dnqbob commented:
And I support circling for Banshee and bomber. 

However, if we really apply ion storm in map,  the needs of immediately landing will be necessary,  which means we need a good circling landing for those aircrafts.

deleted-user-1 commented:
@dnqbob have you tested this pull request?

deleted-user-1 reviewed (dismissed):
This is a great feature and worked as intended when I tested it. Maybe the design decision in this PR can be revised some day in the future.

dnqbob commented:
I feel sorry about it,  but IMO I think a circling Banshee is the way better,  probably because I like Cnc3 aircraft movements.  As well as in CG when Banshee destroy MKII they are just behaving like the normal fixed wings aircrafts. 

Update: 
In other CG when Banshee land on a helipad,  they are not landing verticaly like normal hover aircraft/helicopter. 

But I agree "IdleTurnSpeed" should be applied to all aircrafts that cycling. Making they cycle slowly will help players select them easier and more close to modern CNC. 

Besides,  I am the one like the art and environment of TS but disagree with its balance and some of the game mechanics.  Aircrafts' behavior in idle is one of them. 

In addition,  if I am really that nitpicking I think solve the high pitch of VXLs that will lag the entire game is a more urgent issue compared to current conversation.

deleted-user-1 commented:
> As well as in CG when Banshee destroy MKII they are just behaving like the normal fixed wings aircrafts.

> Update:
> In other CG when Banshee land on a helipad, they are not landing verticaly like normal hover aircraft/helicopter.

Link source or anything? So far you have not provided anything substantial besides your opinion that you like aircraft in c&c3 and don't like ts gameplay. I highly doubt that you actually looked at the changes from this PR or tried the different setups.

You have not brought up a single argument why circling is the better behavior and I find it super frustrating that nevertheless you have had already big influence on key design aspects of the game. Banhees landing during ion storms like you propose will obviously be a horrible hack. :(

deleted-user-1 commented:
@dnqbob 

> I think it is the engine limit make aircrafts not slide/cycling in original TS. because in CG Banshee slides well.

No, that's not what I said. There is a distinction between harpy and banshee, the one can slide and the other can not. Please provide anything to back your implied claim that circling was intended design in ts othwise it's just yet another random opinion. Also no idea what 'CG' is or what you're referring to.

dnqbob commented:
ummm.. I will make some gifs on the CGs when back home. 

CG: mission CG.

deleted-user-1 commented:
> CG: mission CG.

I don't know what that means. Would you have the mercy to explain the abbreviation after I made it clear that I can't read it?

If you're referring to cutscenes, couod you please tell us which one's so everybody can check on their own? I also find cutscene behavior irrelevant when there is clear unit behavior expressed in the game itself (vtol)...

dnqbob commented:
Yeah I mean, cutscenes. Sorry to make you confused about this.

So there we go:
1. MKII get destroyed by Banshees
![s1](https://user-images.githubusercontent.com/13763394/85220460-27cbca00-b3de-11ea-9bd5-8f7c8f44bdbc.gif)

2. Banshees land (not hover)
![s2](https://user-images.githubusercontent.com/13763394/85220467-39ad6d00-b3de-11ea-8e27-83cbd962ae17.gif)

In addition, 
> I also find cutscene behavior irrelevant when there is clear unit behavior expressed in the game itself (vtol)...

So they fix these in C&C3 when they can, even if westwood was dead at that time. And now we can, too.  If we follow the orignal asserts, we should also remove the rotor form Harpy. 

However, this is what Harpy looks like in cutscene:
![85216372-5f268080-b3b6-11ea-9a99-3ebcedba0257](https://user-images.githubusercontent.com/13763394/85220612-98bfb180-b3df-11ea-94cf-978eaaf13f1f.jpg)

Many mods on orignal TS/FS really had enough of the lack of rotor (take a look at TI's GDI Chinook), and obviously it is an incredibly sloppy of Westwood. They were unable to achieve a rotor at that time doesn't means they don't want it, and, IMO, so does the circling & slide when turning like a fixed wings aircraft.

pchote committed:
Allow voxel-based aircraft to pitch and roll.

pchote committed:
Add pitch and roll to TS aircraft.

pchote commented:
Rebased and removed my previous attempts to fix the inconsistencies in bleed's aircraft definitions. This pr is now ONLY about adding the pitch/roll feature. Any further discussion about idle behaviour, fixing the incorrect speeds, or strafing behaviour should go into a different issue.

dnqbob commented:
Agreed.

deleted-user-1 commented:
@dnqbob Thanks for explaining and showing the computer graphics. I assumed you were referring to ts. I agree that the first image implies that banshees indeed circle. The second image can IMO also be interpreted the opposite way, since it shows that the aircraft hovers shortly before landing.

However, this means for me that I can no longer participate in this discusssion since I don't own the later c&c games which do serve nowadays as primary point of orientation when defining openra unit behavior (which I forgot here). So I'd appreciate it a lot if you could repeat your arguments for whatever the behavior in c&c3 is in a new issue so others can then discuss it.

- how is their idle behavior?
- can they move to the cell next to them without flying a curve (sliding)?
- do they have the same behavior as orcas or what are the differences?

Thanks.

pchote commented:
The FMVs show lots of things that do not match gameplay in all of the games, and in many cases (such as the banshees landing like fixed-wing aircraft) *can't* match gameplay.

In any case, this is not the place to have this discussion so please create a new issue.

tovl commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9009b91a7d9d95c401e2a592a53bd950f57802d2...dbe81604f2388c9d5bd3edc606004de21062278e

tovl commented:
@pchote Do you already have follow up PRs in the pipeline for this? Otherwise I could perhaps help out with further (engine level) refinements, like the half-sliding behaviour. But I want to make sure I am not working at cross purposes again :sweat_smile:

pchote commented:
My followup PRs will be focused on turrets and allowing Mobile/Move to tilt on slopes. Aircraft sliding is all yours.
